src/tree/updater_quantile_hist.cc

```diff
@@ -152,26 +203,51 @@ void QuantileHistMaker::Builder::BuildHistogramsLossGuide(
   int starting_index = std::numeric_limits<int>::max();
   int sync_count = 0;
 
-  AddHistRows(&starting_index, &sync_count);
+  AddHistRows(&starting_index, &sync_count, p_tree);
   BuildLocalHistograms(gmat, gmatb, p_tree, gpair_h);
   SyncHistograms(starting_index, sync_count, p_tree);
 }
 
 
-void QuantileHistMaker::Builder::AddHistRows(int *starting_index, int *sync_count) {
+void QuantileHistMaker::Builder::AddHistRows(int *starting_index, int *sync_count,
+                                             RegTree *p_tree) {
   builder_monitor_.Start("AddHistRows");
 
-  for (auto const& entry : nodes_for_explicit_hist_build_) {
-    int nid = entry.nid;
-    hist_.AddHistRow(nid);
-    (*starting_index) = std::min(nid, (*starting_index));
+  std::vector<int> merged_hist(nodes_for_explicit_hist_build_.size() +
+                                  nodes_for_subtraction_trick_.size());
+  for (size_t i = 0; i < nodes_for_explicit_hist_build_.size(); ++i) {
+    merged_hist[i] = nodes_for_explicit_hist_build_[i].nid;
   }
-  (*sync_count) = nodes_for_explicit_hist_build_.size();
-
-  for (auto const& node : nodes_for_subtraction_trick_) {
-    hist_.AddHistRow(node.nid);
+  for (size_t i = 0; i < nodes_for_subtraction_trick_.size(); ++i) {
+    merged_hist[nodes_for_explicit_hist_build_.size() + i] =
+    nodes_for_subtraction_trick_[i].nid;
+  }
+  std::sort(merged_hist.begin(), merged_hist.end());
+  int n_left = 0;
+  for (auto const& nid : merged_hist) {
+    if ((*p_tree)[nid].IsLeftChild()) {
+      hist_.AddHistRow(nid);
+      (*starting_index) = std::min(nid, (*starting_index));
+      n_left++;
+      if (rabit::IsDistributed()) {
+        phist_local_.AddHistRow(nid);
+      }
+    }
+  }
+  for (auto const& nid : merged_hist) {
+    if (!((*p_tree)[nid].IsLeftChild())) {
+      hist_.AddHistRow(nid);
+      if (rabit::IsDistributed()) {
+        phist_local_.AddHistRow(nid);
+      }
+    }
   }
 
+  (*sync_count) = merged_hist.size() / 2;
+  if (*sync_count == 0) {
+    (*sync_count) = 1;
+  }
+  CHECK_EQ(n_left, (*sync_count));
```

Comment:
It's equal to # of left nodes or 1 for the root. So, it can be simplified.

Reply:
Also, I'm not sure that a condition in the check will be always true.

Reply:
simplified